Tom Perkins exercises - ex 2 (shapes)

tom-perkins opened the pull request:
WordCounter exercise: counts number of occurrences of unique words in an ascii file

Commit:
First commit - create exercises files

Created exercises files as a copy of the template

Commit:
WordCounter: accept filename as an argument

WordCounter is a command-line program that takes the filename of a text
file as an argument.
If no file is specified, a usage message is displayed.

Commit:
WordCounter: Counter class for counting words

The Counter class is an associative array for keeping track of how many
times each word has been seen

Commit:
WordCounter: reads from file and counts words

First attempt at functionality. Still to do:
- only count words > 4 characters
- deal with case and punctuation, hyphens
- sort output list
- put a header on output list

Commit:
WordCounter: Sort output list by word frequency

The output table of word versus frequency should be in descending order
of usage.

Commit:
WordCounter: validation of input words

Strip punctuation, case-insensitive, only consider words of more than
four characters.

Commit:
WordCounter: specify output file as argument

Output file is specified as a command-line parameter instead of writing
to stdout

Commit:
WordCounter: add comment to main

Improve comments on main to describe what the program does

OwenArnold commented on exercises/tom_perkins/ex01_basics/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+
+set ( SRC_FILES
+  src/main.cpp
+  # add any additional source files here
```
Build is currently failing because `src/counter.cpp` is missing

OwenArnold commented on exercises/tom_perkins/ex01_basics/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+)
+
+set ( HDR_FILES
+  # add any additional header files here
```
Also need `src/counter.h` added

Commit:
WordCounter: add files to CMake file

Build failed on Jenkins, possibly because the CMakeLists.txt didn't
include the new files. Add them and try building again.

OwenArnold commented on exercises/tom_perkins/ex01_basics/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+  # add any additional header files here
+)
+
+add_executable ( WordCounter ${SRC_FILES} ${HDR_FILES} )
```
I got a warning on my platform because you're using c++11 features without the flags being set correctly. You can fix this by adding the following.
`set_property(TARGET WordCounter PROPERTY CXX_STANDARD 11)` **However** the RHEL6 build will continue to fail because it's using an older version of GCC, which won't support features such as the range-based for loops you used.

OwenArnold commented on exercises/tom_perkins/ex01_basics/src/counter.h:
```diff
@@ -0,0 +1,31 @@
+ * and can report on how many times it has seen each.
+ * Currently works for ASCII only (char not wchar_t).
+ */
+class Counter
```
Use of separate well-named class for this is good. Not that it's wrong, but why did you use a class over say a free function?

OwenArnold commented on exercises/tom_perkins/ex01_basics/src/counter.cpp:
```diff
@@ -0,0 +1,87 @@
+#include "counter.h"
+
+/**
+ * Compares two (word, frequency) pairs to enable sorting.
```
Good use of comments and free-function. But better to put in an anonymous namespace to keep linkage internal. Good use of const refs.

OwenArnold commented on exercises/tom_perkins/ex01_basics/src/counter.cpp:
```diff
@@ -0,0 +1,87 @@
+ * Writes a table of words and frequencies to supplied stream.
+ * Words are sorted in descending order of usage.
+ */
+void Counter::generateReport(std::ostream& stream)
```
Good function naming. 
- Prefer to return something rather than modify the stream directly (I'll explain in detail in a bit)
- Shouldn't this function be const?
- What about error handling considerations?

Commit:
WordCounter: code review

- Remove unsupported C++11 features
- Counter::generateReport is a const member function
- Put free functions in an anonymous namespace

Commit:
Shapes: create files and add to CMakeLists

Create new files for project and add to the list for CMake so the build
will work

OwenArnold commented:
Looks good so far. Moving forward. We use clang-format for code formatting. There's a VS plugin for that. Also we use doxygen for code comments.

Commit:
Shapes: abstract base class

Shape is an abstract base class representing a shape

Commit:
Shape: implement Circle class

Circle is a class to represent a circle, derived from Shape

Commit:
Shape: implement Rectangle, Square and Triangle

Square is a special case of Rectangle where the two sides are the same.

Commit:
Shape: create a vector of shapes

Make sure that all derived shapes inherit the public interface to Shape.
This enables the polymorphism used when constructing a vector of (smart
pointers to) Shapes.

Edit main to test that the vector works as intended.

Commit:
Shape: start implementation of ShapeSorter

ShapeSorter selects strings based on their type, number of sides, or
sorts strings by perimeter or area.

Commit:
Shape: give each shape a name

The ShapeSorter is supposed to print a list of the shapes, sorted or
filtered as appropriate. This isn't very meaningful if the shapes aren't
called anything.

Give each shape a name by which it will be referred to.
Then the ShapeSorter can print a list of shapes' names and the output
will be more meaningful.

Commit:
Shape: implement selecting by number of sides

ShapeSorter can now select shapes by number of sides.
When selecting by type, it is also now case-insensitive.

Commit:
Shape: change ShapeSorter to return sorted vector

ShapeSorter should return the sorted/selected collection of shapes, not
print to screen itself.

Commit:
Shape: make ShapeSorter::select generic

Use a ShapeTester to test whether a shape should be selected or not. For
the case of selecting shapes of a certain type, use the derived
TypeTester.

Commit:
Shape: enable selecting shapes with n sides

SidesTester is a ShapeTester that tests for a certain number of sides

Commit:
Shape: ShapeSorter can sort by area or perimeter

Implement functionality for ShapeSorter to sort the vector of shapes by
area or by perimeter.

Commit:
Shape: tidy up main by refactoring print into a function

Printing a vector of shapes to screen was done in several places, so has
been refactored into a utility function.

Commit:
Shape: capitalisation of include file

Commit:
Shape: Include cmath for triangle as it needs sqrt

Commit:
Shape: ShapeTester::match takes a reference

Changed ShapeTester::match to take a const Shape& rather than a const
shared_ptr<Shape>&

Commit:
Shape: remove unnecessary includes

TypeTester and ShapeSorter don't need to include <string> because they
already include a header that does

Commit:
Shape: apply clang formatting to all files

Used visual studio clang-format plugin to clean up whitespace and code
formatting in all files. No changes to code in this commit.

Commit:
Shape: sorting should be from largest to smallest

Reverse the order of sorting for area/perimeter - requirement is to sort
from largest to smallest

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+  src/circle.h 
+  src/triangle.h   
+  src/shapesorter.h 
+  src/ShapeTester.h
```
File naming case is differnt. On Mantid we use Pascal case. Except for python files.

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/ShapeTester.h:
```diff
@@ -0,0 +1,18 @@
+ */
+class ShapeTester {
+public:
+  virtual bool match(const Shape &toCompare) const = 0;
```
Excellent

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/shape.h:
```diff
@@ -0,0 +1,29 @@
+  virtual double area() const = 0;
+  const int sides;
+  const std::string type;
+  std::string name;
```
Why is name non-const but the others are const?

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/shapesorter.cpp:
```diff
@@ -0,0 +1,119 @@
+#include "shapesorter.h"
+
+typedef std::vector<std::shared_ptr<Shape>>
```
Nicely done. I would probably also typedef std::shape_ptr as something like Shape_sprt

Also, I believe that your Shapes are immutable (if done correctly). So you could store them as std::shared_ptr<const Shape>

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/main.cpp:
```diff
@@ -0,0 +1,58 @@
+int main(int argc, char **argv) {
+  // Create shapes and put them in a vector
+  std::vector<std::shared_ptr<Shape>> shapes;
+  shapes.push_back(std::make_shared<Circle>(1.0, "circle1"));
```
Good

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/rectangle.h:
```diff
@@ -0,0 +1,24 @@
+  virtual double area() const;
+
+private:
+  double x;
```
should be `const double`

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/rectangle.cpp:
```diff
@@ -0,0 +1,46 @@
+ */
+Rectangle::Rectangle(double xIn, double yIn, std::string nameIn)
+    : Shape(4, "rectangle", nameIn) {
+  x = xIn;
```
I think you should use an initialiser list here instead of assignment

OwenArnold commented on exercises/tom_perkins/ex02_oo_basics/src/rectangle.cpp:
```diff
@@ -0,0 +1,46 @@
+ * \brief Calculates the perimeter of this rectangle.
+ * \returns Perimeter of the rectangle.
+ */
+double Rectangle::perimeter() const { return x + x + y + y; }
```
Good

Commit:
Shape: const-correctness

Shapes should be immutable - can't change their dimensions once created.
Use initializer lists in constructors.
